File: java/factorial_calculator/factorial_calculator_4.java
/**
 * This program is a factorial calculator written in Java.
 * It calculates the factorial of a given number using a loop.
 * The purpose of this program is to demonstrate basic programming concepts
 * such as loops, conditionals, and user input handling.
 * The program will prompt the user to enter a non-negative integer and then
 * calculate and display the factorial of that number.
 */

import java.util.Scanner;

public class FactorialCalculator {

    public static void main(String[] args) {
        // Create a Scanner object to read input from the user
        Scanner scanner = new Scanner(System.in);

        // Prompt the user to enter a non-negative integer
        System.out.print("Enter a non-negative integer: ");
        int number = scanner.nextInt();

        // Check if the entered number is non-negative
        if (number < 0) {
            System.out.println("Please enter a non-negative integer.");
        } else {
            // Calculate the factorial of the entered number
            long factorial = calculateFactorial(number);

            // Display the result
            System.out.println("The factorial of " + number + " is " + factorial);
        }

        // Close the scanner
        scanner.close();
    }

    /**
     * This method calculates the factorial of a given number using a loop.
     * 
     * @param n the number to calculate the factorial of
     * @return the factorial of the given number
     */
    public static long calculateFactorial(int n) {
        long result = 1;

        // Loop to calculate the factorial
        for (int i = 1; i <= n; i++) {
            result *= i;
        }

        // Return the calculated factorial
        return result;
    }
}

File: perl/factorial_calculator/factorial_calculator_4.pl
#!/usr/bin/perl
use strict;
use warnings;

# This program calculates the factorial of a given number.
# The purpose of this program is to help programmers understand
# how to implement a basic factorial calculator in Perl.
# The program will prompt the user to enter a number and then
# calculate the factorial of that number using a recursive function.

# Function to calculate factorial recursively
sub factorial {
    my ($n) = @_;
    
    # Base case: factorial of 0 or 1 is 1
    if ($n == 0 || $n == 1) {
        return 1;
    }
    
    # Recursive case: n * factorial of (n-1)
    return $n * factorial($n - 1);
}

# Main program
print "Enter a number to calculate its factorial: ";
my $number = <STDIN>;
chomp($number);

# Input validation: ensure the input is a non-negative integer
if ($number =~ /^\d+$/ && $number >= 0) {
    my $result = factorial($number);
    print "The factorial of $number is $result\n";
} else {
    print "Please enter a valid non-negative integer.\n";
}

